File: packages/zod/src/ebook.ts
import { z } from 'zod'
import { ZModel } from './utils.js'

export const ZEbookFormat = z.enum(['epub', 'pdf', 'txt'])

export const ZEbook =
	z
		.object({
			ownerUserId: z.string().uuid(),
			title: z.string(),
			description: z.string().optional(),
			format: ZEbookFormat,
			languageCode: z.string().optional(),
			storageKey: z.string(),
			fileSizeBytes: z.number().int(),
			checksumSha256: z.string(),
			importedAt: z.string().datetime(),
		})
		.extend(ZModel.shape)

export const ZStoreEbookDTO = z.object({
	title: z.string().min(1).max(255),
	description: z.string().optional(),
	format: ZEbookFormat,
	languageCode: z.string().max(16).optional(),
	storageKey: z.string().min(1),
	fileSizeBytes: z.number().int().positive(),
	checksumSha256: z.string().length(64),
	importedAt: z.string().datetime().optional(),
})

export const ZUpdateEbookDTO = z.object({
	title: z.string().min(1).max(255).optional(),
	description: z.string().optional(),
	languageCode: z.string().max(16).optional(),
})

export const ZEbookGoogleMetadata = z.object({
	ebookId: z.string().uuid(),
	googleBooksId: z.string(),
	isbn10: z.string().optional(),
	isbn13: z.string().optional(),
	publisher: z.string().optional(),
	publishedDate: z.string().optional(),
	pageCount: z.number().int().optional(),
	categories: z.array(z.string()).optional(),
	thumbnailUrl: z.string().url().optional(),
	infoLink: z.string().url().optional(),
	rawPayload: z.record(z.any()).optional(),
	attachedAt: z.string().datetime(),
	updatedAt: z.string().datetime(),
	deletedAt: z.string().datetime().optional(),
})

export const ZAttachGoogleMetadataDTO = z.object({
	googleBooksId: z.string().min(1),
	isbn10: z.string().length(10).optional(),
	isbn13: z.string().length(13).optional(),
	publisher: z.string().optional(),
	publishedDate: z.string().optional(),
	pageCount: z.number().int().positive().optional(),
	categories: z.array(z.string()).optional(),
	thumbnailUrl: z.string().url().optional(),
	infoLink: z.string().url().optional(),
	rawPayload: z.record(z.any()).optional(),
})

File: apps/api/internal/application/dto/ebook.go
package dto

import (
	"time"

	"github.com/google/uuid"
	"github.com/jeheskielSunloy77/libra-link/internal/domain"
)

type StoreEbookInput struct {
	OwnerUserID    uuid.UUID
	Title          string
	Description    *string
	Format         domain.EbookFormat
	LanguageCode   *string
	StorageKey     string
	FileSizeBytes  int64
	ChecksumSHA256 string
	ImportedAt     *time.Time
}

func (d *StoreEbookInput) ToModel() *domain.Ebook {
	importedAt := time.Now().UTC()
	if d.ImportedAt != nil {
		importedAt = d.ImportedAt.UTC()
	}

	return &domain.Ebook{
		OwnerUserID:    d.OwnerUserID,
		Title:          d.Title,
		Description:    d.Description,
		Format:         d.Format,
		LanguageCode:   d.LanguageCode,
		StorageKey:     d.StorageKey,
		FileSizeBytes:  d.FileSizeBytes,
		ChecksumSHA256: d.ChecksumSHA256,
		ImportedAt:     importedAt,
	}
}

type UpdateEbookInput struct {
	Title        *string
	Description  *string
	LanguageCode *string
}

func (d *UpdateEbookInput) ToModel() *domain.Ebook {
	out := &domain.Ebook{}
	if d.Title != nil {
		out.Title = *d.Title
	}
	if d.Description != nil {
		out.Description = d.Description
	}
	if d.LanguageCode != nil {
		out.LanguageCode = d.LanguageCode
	}
	return out
}

func (d *UpdateEbookInput) ToMap() map[string]any {
	updates := map[string]any{}
	if d.Title != nil {
		updates["title"] = *d.Title
	}
	if d.Description != nil {
		updates["description"] = d.Description
	}
	if d.LanguageCode != nil {
		updates["language_code"] = d.LanguageCode
	}
	return updates
}

type AttachGoogleMetadataInput struct {
	EbookID       uuid.UUID
	GoogleBooksID string
	ISBN10        *string
	ISBN13        *string
	Publisher     *string
	PublishedDate *string
	PageCount     *int
	Categories    []string
	ThumbnailURL  *string
	InfoLink      *string
	RawPayload    map[string]any
}
